FHEM/00_SIGNALduino.pm

```diff
@@ -1329,7 +1341,11 @@ sub SIGNALduino_CheckVersionResp
 		$hash->{keepalive}{ok}    = 0;
 		$hash->{keepalive}{retry} = 0;
 		InternalTimer(gettimeofday() + SDUINO_KEEPALIVE_TIMEOUT, "SIGNALduino_KeepAlive", $hash, 0);
-	 	$hash->{cc1101_available} = 1  if ($hash->{version} =~ m/cc1101/);
+		if ($hash->{version} =~ m/cc1101/) {
+			$hash->{cc1101_available} = 1;
+			InternalTimer(gettimeofday() + SDUINO_GET_CONFIGQUERY_DELAY, "SIGNALduino_delayed_Get", "SIGNALduino_delayed_Get:$name:ccconf", 0);
+			InternalTimer(gettimeofday() + SDUINO_GET_CONFIGQUERY_DELAY + 1, "SIGNALduino_delayed_Get", "SIGNALduino_delayed_Get:$name:ccpatable", 0);
```

Comment:
Wieso muss hier eine Zusätzliche Sekunde gewartet werden?

Reply:
Weil der Ablauf sonst gestört wurde. 
Es wurde sonst keine Einstellung gelesen als ob etwas überschrieben wurde.

Reply:
Sorry, meine Frage war unpräzise und vermutlich nerft es auch ein bisschen, aber ich möchte das gerne Verstehen.

Du hast SDUINO_GET_CONFIGQUERY_DELAY mit 0.75 Sekunden definiert.
Diese Zeit wird verwendet um einen GET Befehl zu verzögern, in der Hoffnung, dass bis dahin die Antwort vom vorherigen get Befehl empfangen wurde.

Wenn ein weiterer Getbefehl verzögert werden soll, wird aber Eine Sekunde gewartet.
Mich irritieren die unterschiedlichen Werte, die von verschiedenen Faktoren abhängen können.
Wenn der Anwender zwischendurch noch einen anderen Befehl absendet, ist das meiner Einschätzung nach ohnehin nicht mehr funktional.

Ich hätte da zwei Ansätze, die es ein wenig anders lösen würden.

a) In SIGNALduino_delayed_Get wird überprüft ob noch auf Antwort eine Getbefehls gewartet wird. Ist noch einer ausstehend, wird erneut verzögert, andernfalls wird der Befehl abgesetzt.
Dabei kann die Reihenfolge der Befehle durcheinander geraten, was auf den Ersten Blick aber kein Problem zu sein scheint.

b) Wir bauen die Verarbeitung der Get Antworten so um, dass auch mehrere ausstehende Antworten verarbeitet werden können. 


In beiden Fällen, würde immer nur dann ein neuer Befehl gesendet, wenn die Antwort vom vorherigen verarbeitet wurde, egal wie lange es dauert.
GGf. müsste man das noch mit einem max timeout vor der Unendlichkeit absichern :)

Reply:
Wäre es denn nicht viel einfacher, wieder alle Nachrichten, die vom SIGNALduino kommen, zu parsen?
Ich weiß, du hast etwas gegen lange if-Verschachtelungen, aber das spart den ganzen Aufwand mit irgendwelchen Timern und Übergaben von den Subs die es verarbeiten sollen.

Reply:
Was meinst Du denn mit "alle Nachrichten" parsen?

Reply:
> Sorry, meine Frage war unpräzise und vermutlich nerft es auch ein bisschen, aber ich möchte das gerne Verstehen.
> 
> Du hast SDUINO_GET_CONFIGQUERY_DELAY mit 0.75 Sekunden definiert.
> Diese Zeit wird verwendet um einen GET Befehl zu verzögern, in der Hoffnung, dass bis dahin die Antwort vom vorherigen get Befehl empfangen wurde.
> 
> Wenn ein weiterer Getbefehl verzögert werden soll, wird aber Eine Sekunde gewartet.
> Mich irritieren die unterschiedlichen Werte, die von verschiedenen Faktoren abhängen können.
> Wenn der Anwender zwischendurch noch einen anderen Befehl absendet, ist das meiner Einschätzung nach ohnehin nicht mehr funktional.
> 
> Ich hätte da zwei Ansätze, die es ein wenig anders lösen würden.
> 
> a) In SIGNALduino_delayed_Get wird überprüft ob noch auf Antwort eine Getbefehls gewartet wird. Ist noch einer ausstehend, wird erneut verzögert, andernfalls wird der Befehl abgesetzt.
> Dabei kann die Reihenfolge der Befehle durcheinander geraten, was auf den Ersten Blick aber kein Problem zu sein scheint.
> 
> b) Wir bauen die Verarbeitung der Get Antworten so um, dass auch mehrere ausstehende Antworten verarbeitet werden können.
> 
> In beiden Fällen, würde immer nur dann ein neuer Befehl gesendet, wenn die Antwort vom vorherigen verarbeitet wurde, egal wie lange es dauert.
> GGf. müsste man das noch mit einem max timeout vor der Unendlichkeit absichern :)

Ich kann mir schon vorstellen, das du es nachvollziehen möchtest. Teste es doch mal live bei dir aus. So verstehst du es am eigenen Gerät am besten.

Die genannten Varianten sind möglich. Welche Variante würdest du favorisieren oder welche Variante wäre nicht zu aufwendig ?

Nervig, was ist das? Jeder möchte nur verstehen :)

Reply:
Ich meinte, wenn alle Nachrichten, die vom SIGNALduino kommen, in einer Sub verarbeitet werden, dürfte es solche Schwierigkeiten gar nicht erst geben.

Reply:
Die Nachrichten werden alle verarbeitet.
Leider findet immer nur eine Nachricht die passende Zuordnung.
Letzteres geht uns verloren, wenn ein neuer Befehl abgesetzt wird.

Grundsätzlich bringt mich das aber gerade auf eine Idee, dass wir verhindern könnten, dass ein neuer Befehl einen stehenden überschreibt.
Das wäre dann auch über die GUI nicht mehr möglich.

Vermutlich hast Du aber was ganz anderes im Sinn gehabt :)

Reply:
Kurzgefasst,
Letztendlich geht es doch nicht unbedingt um das überschreiben. Es geht doch um das „Ende“ von einer Befehlsfolge und eine neue vom GUI abzusenden.

Das kann man sicherlich auch erzielen wenn eine Warteschlange wie beim uC vorhanden wäre.

Wir sollten uns die Frage stellen, was ist am effektivsten für die Zukunft und was ist ebenso verständlich nach zu vollziehen. Danach sollten wir handeln :)